src/set_theory/ordinal_arithmetic.lean

```diff
@@ -894,38 +943,310 @@ match o, o.out, o.out_eq, f :
+bsup_eq_sup _
+
+theorem blsub_le_iff_lt {o f a} : blsub o f ≤ a ↔ ∀ i h, f i h < a :=
+by { convert bsup_le, apply propext, simp [succ_le] }
+
+theorem lt_blsub {o} (f : Π a < o, ordinal) (i h) : f i h < blsub o f :=
+blsub_le_iff_lt.1 (le_refl _) _ _
+
+theorem bsup_le_blsub {o} (f : Π a < o, ordinal) : bsup o f ≤ blsub o f :=
+bsup_le.2 (λ i h, le_of_lt (lt_blsub f i h))
+
+theorem blsub_le_bsup_succ {o} (f : Π a < o, ordinal) : blsub o f ≤ (bsup o f).succ :=
+blsub_le_iff_lt.2 (λ i h, lt_succ.2 (le_bsup f i h))
+
+theorem bsup_succ_le_blsub {o} (f : Π a < o, ordinal) :
+  (bsup o f).succ ≤ blsub o f ↔ ∃ i hi, f i hi = bsup o f :=
+begin
+  refine ⟨λ h, _, _⟩,
+  { by_contra' hf,
+    exact ne_of_lt (succ_le.1 h) (le_antisymm (bsup_le_blsub f)
+      (blsub_le_iff_lt.2 (lt_bsup_of_ne_bsup.1 hf))) },
+  rintro ⟨_, _, hf⟩,
+  rw [succ_le, ←hf],
+  exact lt_blsub _ _ _
+end
+
+theorem bsup_succ_eq_blsub {o} (f : Π a < o, ordinal) :
+  (bsup o f).succ = blsub o f ↔ ∃ i hi, f i hi = bsup o f :=
+begin
+  rw iff.intro le_of_eq (λ h, le_antisymm h (blsub_le_bsup_succ f)),
+  exact bsup_succ_le_blsub f
+end
+
+theorem bsup_eq_blsub {o} (f : Π a < o, ordinal) :
+  bsup o f = blsub o f ↔ ∀ a < blsub o f, succ a < blsub o f :=
+by { rw [bsup_eq_sup, blsub_eq_lsub], exact sup_eq_lsub _ }
+
+theorem blsub_eq_zero_iff {o} {f : Π a < o, ordinal} : blsub o f = 0 ↔ o = 0 :=
+by { rw [blsub_eq_lsub, lsub_eq_zero_iff], exact out_empty_iff_eq_zero }
+
+theorem blsub_type (r : α → α → Prop) [is_well_order α r] (f) :
+  blsub (type r) f = lsub (λ a, f (typein r a) (typein_lt_type _ _)) :=
+eq_of_forall_ge_iff $ λ o,
+by rw [blsub_le_iff_lt, lsub_le_iff_lt]; exact
+  ⟨λ H b, H _ _, λ H i h, by simpa only [typein_enum] using H (enum r i h)⟩
+
+theorem blsub_id {o} : blsub.{u u} o (λ x _, x) = o :=
+begin
+  apply le_antisymm,
+  { rw blsub_le_iff_lt,
+    exact λ _, id },
+  by_contra' h,
+  exact lt_irrefl _ (lt_blsub.{u u} (λ x _, x) _ h)
+end
+
+/-! ### Enumerating unbounded sets of ordinals with ordinals -/
+
+section
+variables {S : set ordinal.{u}} (hS : unbounded (<) S)
+
+-- A characterization of unboundedness that's more convenient to our purposes.
+private lemma unbounded_aux (hS : unbounded (<) S) (a) : ∃ b, b ∈ S ∧ a ≤ b :=
+by { rcases hS a with ⟨b, hb, hb'⟩, exact ⟨b, hb, le_of_not_gt hb'⟩ }
+
+/-- Enumerator function for an unbounded set of ordinals. -/
+def enum_ord : ordinal.{u} → ordinal.{u} :=
+wf.fix (λ o f, omin _ (unbounded_aux hS (blsub.{u u} o f)))
+
+/-- The hypothesis that asserts that the `omin` from `enum_ord_def'` exists. -/
+lemma enum_ord_def'_H {hS : unbounded (<) S} {o} :
+  ∃ x, x ∈ S ∧ blsub.{u u} o (λ c _, enum_ord hS c) ≤ x :=
+unbounded_aux hS _
+
+/-- The equation that characterizes `enum_ord` definitionally. This isn't the nicest expression to
+work with, so consider using `enum_ord_def` instead. -/
+theorem enum_ord_def' (o) :
+  enum_ord hS o = omin (λ b, b ∈ S ∧ blsub.{u u} o (λ c _, enum_ord hS c) ≤ b) enum_ord_def'_H :=
```

Comment:
The first argument of `omin` is a `set ordinal`, so this should be using set notation `{b | ...}` .

Reply:
Fixed in #11369.